### PR TITLE
Refactoring/gcp/cloudsql

### PR DESCRIPTION
**Pull request by Remi05**
Migrated GCP's Cloud SQL service to the new architecture.

Note that `gcp/resources/projects.py` and `gcp/facade/base.py` are the same as in the [Stackdriver logging PR](https://github.com/nccgroup/ScoutSuite/pull/229) and the [IAM PR](https://github.com/nccgroup/ScoutSuite/pull/230); they are simply shared by all services and will appear in all GCP refactoring PRs until one of the services is merged.

### CODE REVIEW

**Comment by j4v** [comment]
:clap: :clap: :clap:

File: ScoutSuite/__main__.py [hunk 1]
```diff
@@ -62,17 +61,12 @@
...
-                                                    user_account=args.get(
-                                                        'user_account'),
-                                                    service_account=args.get(
-                                                        'service_account'),
+        authenticated = cloud_provider.authenticate(profile=args.get('profile'),
+                                                    csv_credentials=args.get('csv_credentials'),
+                                                    mfa_serial=args.get('mfa_serial'),
+                                                    mfa_code=args.get('mfa_code'),
+                                                    user_account=args.get('user_account'),
+                                                    service_account=args.get('service_account'),
```
Comment (Aboisier) [inline ScoutSuite/__main__.py]: We keep reverting that change, can we just leave it like it was haha

**Comment by misg** [comment]
Could you please merge with `refactoring/gcp-resource-configs` to avoid the 263 commits and 86 files changed (I can't do my review in those conditions, sorry :sob:).

File: ScoutSuite/providers/gcp/facade/utils.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+            resources.extend(response.get(resource_key, []))
+            request = resources_group.list_next(previous_request=request, previous_response=response)
+
+    @staticmethod
+    async def get_all(resource_key, request, resources_group):
+        resources = []
+        await run_concurrently(
+            lambda: GCPFacadeUtils._get_all(resources, resource_key, request, resources_group)
+        )
+        return resources
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/facade/utils.py]: I'm nitpicking, but I think it makes more sense to put `get_all` before `_get_all` (when you read the file, you start from a high level and you go deeper, it makes it more readable IMO).

File: ScoutSuite/providers/gcp/facade/utils.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import httplib2
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/facade/utils.py]: Unused import

File: ScoutSuite/providers/gcp/facade/utils.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import googleapiclient
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/facade/utils.py]: Unused import

File: ScoutSuite/providers/gcp/facade/utils.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+from ScoutSuite.providers.utils import run_concurrently
+
+class GCPFacadeUtils:
+    @staticmethod
+    def _get_all(resources, resource_key, request, resources_group):
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/facade/utils.py]: Are there  parameters you can add type hints to?

File: ScoutSuite/providers/gcp/facade/utils.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+        while request is not None:
+            response = request.execute()
+            resources.extend(response.get(resource_key, []))
+            request = resources_group.list_next(previous_request=request, previous_response=response)
+
+    @staticmethod
+    async def get_all(resource_key, request, resources_group):
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/facade/utils.py]: Same here for the type hints

File: ScoutSuite/providers/gcp/facade/utils.py (hunk 5 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/facade/utils.py]: I didn't find any types for `resources`, `request` and `resources_group` as the library is auto-generated using the Google Cloud API description.

**Comment by Remi05** [comment]
Fixed UI and added left menu for projects.

cloudsql_projects

File: ScoutSuite/providers/gcp/facade/cloudsql.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+class CloudSQLFacade(GCPBaseFacade):
+    def __init__(self):
+        super(CloudSQLFacade, self).__init__('sqladmin', 'v1beta4')
+
+    async def get_backups(self, project_id: str, instance_name: str):
+        cloudsql_client = self._get_client()
+        backups_group = cloudsql_client.backupRuns()
+        request = backups_group.list(project=project_id, instance=instance_name)
+        return await GCPFacadeUtils.get_all('items', request, backups_group)
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/facade/cloudsql.py]: Shouldn't this be parallelized?

File: ScoutSuite/providers/gcp/facade/cloudsql.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+    async def get_database_instances(self, project_id: str):
+        cloudsql_client = self._get_client()
+        instances_group = cloudsql_client.instances()
+        request = instances_group.list(project=project_id)
+        return await GCPFacadeUtils.get_all('items', request, instances_group)
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/facade/cloudsql.py]: Same here, shouldn't this be parallelized?

File: ScoutSuite/providers/gcp/facade/cloudsql.py (hunk 7 above)
Comment (misg) [reply ScoutSuite/providers/gcp/facade/cloudsql.py]: It's already parallelized in the `get_all` method.

File: ScoutSuite/providers/gcp/resources/cloudsql/backups.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+class Backups(Resources):
+    def __init__(self, gcp_facade: GCPFacade, project_id: str, instance_name: str):
+        self.gcp_facade = gcp_facade
+        self.project_id = project_id
+        self.instance_name = instance_name
+
+    async def fetch_all(self):
+        raw_backups = await self.gcp_facade.cloudsql.get_backups(self.project_id, self.instance_name)
+        for raw_backup in raw_backups:
+            if raw_backup['status'] == 'SUCCESSFUL':
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/resources/cloudsql/backups.py]: I guess it was like that before, but what about the failed backups?

File: ScoutSuite/providers/gcp/resources/cloudsql/database_instances.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    def __init__(self, gcp_facade: GCPFacade, project_id: str):
+        self.gcp_facade = gcp_facade
+        self.project_id = project_id
+
+    async def fetch_all(self):
+        raw_instances = await self.gcp_facade.cloudsql.get_database_instances(self.project_id)
+        for raw_instance in raw_instances:
+            instance_id, instance = self._parse_instance(raw_instance)
+            self[instance_id] = instance
+            await self._fetch_children(self[instance_id], gcp_facade = self.gcp_facade, project_id = self.project_id, instance_name = instance['name'])
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/resources/cloudsql/database_instances.py]: I believe this should be parallelized using `asyncio.ensure_future`.

File: ScoutSuite/providers/gcp/facade/cloudsql.py (hunk 7 above)
Comment (Aboisier) [reply ScoutSuite/providers/gcp/facade/cloudsql.py]: Oh my bad, I looked at the `_get_all`.

File: ScoutSuite/providers/gcp/resources/cloudsql/database_instances.py (hunk 10 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/resources/cloudsql/database_instances.py]: Done!

File: ScoutSuite/providers/gcp/resources/cloudsql/backups.py (hunk 9 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/resources/cloudsql/backups.py]: I guess they are considered not to be relevant, but @j4v would know more!

**Pull request merged**